sgkit/io/vcfzarr_reader.py

```diff
@@ -0,0 +1,109 @@
+from typing import Any, List, Tuple
+
+import dask.array as da
+import numpy as np
+import xarray as xr
+import zarr
+
+from ..api import DIM_VARIANT, create_genotype_call_dataset
+from ..typing import ArrayLike, PathType
+
+
+def encode_array(x: ArrayLike) -> Tuple[ArrayLike, List[str]]:
+    """Encode array values as integers indexing unique values
+
+    The codes created for each unique element in the array correspond
+    to order of appearance, not the natural sort order for the array
+    dtype.
+
+    Examples
+    --------
+
+    >>> encode_array(['c', 'a', 'a', 'b'])
+    (array([0, 1, 1, 2]), array(['c', 'a', 'b'], dtype='<U1'))
+
+    Parameters
+    ----------
+    x : (M,) array-like
+        Array of elements to encode of any type
+
+    Returns
+    -------
+    indexes : (M,) ndarray
+        Encoded values as integer indices
+    values : ndarray
+        Unique values in original array in order of appearance
+    """
+    # argsort not implemented in dask: https://github.com/dask/dask/issues/4368
+    names, index, inverse = np.unique(x, return_index=True, return_inverse=True)
+    index = np.argsort(index)
+    rank = np.empty_like(index)
+    rank[index] = np.arange(len(index))
+    return rank[inverse], names[index]
+
+
+def read_vcfzarr(path: PathType) -> xr.Dataset:
+    """Read a VCF Zarr file.
+
+    Loads VCF variant, sample, and genotype data as Dask arrays within a Dataset
+    from a Zarr file created using scikit-allel's `vcf_to_zarr` function.
+
+    Since `vcf_to_zarr` does not preserve phasing information, there is no
+    `call/genotype_phased` variable in the resulting dataset.
+
+    Parameters
+    ----------
+    path : PathType
+        Path to the Zarr file.
+
+    Returns
+    -------
+    xr.Dataset
+        The dataset of genotype calls, created using `create_genotype_call_dataset`.
+    """
+
+    vcfzarr = zarr.open_group(str(path), mode="r")
```

Comment:
```suggestion
def read_vcfzarr(store: Union[str, Mapping], consolidated=False: bool) -> xr.Dataset:
    """Read data in Zarr format that was created using scikit-allel's `vcf_to_zarr` function.

    Loads variant, sample, and genotype data as Dask arrays within a Dataset.

    Since `vcf_to_zarr` does not preserve phasing information, there is no
    `call/genotype_phased` variable in the resulting dataset.

    Parameters
    ----------
    store : Union[str, Mapping]
        Either a string providing a URL or local file system path, or a store object.
    consolidated : bool
        Set True if the store uses consolidated metadata.
    
    Returns
    -------
    xr.Dataset
        The dataset of genotype calls, created using `create_genotype_call_dataset`.
    """

    if isinstance(path, str):
        # assume path is an fsspec-style url path
        import fsspec
        store = fsspec.get_mapper(path)
    else:
        # assume path is a store
        store = path
    if consolidated:
        vcfzarr = zarr.open_consolidated(store=store, mode="r")
    else:
        vcfzarr = zarr.open(store=store, mode="r")
```

Two suggestions here.

First suggestion is to allow the first argument to be either a string or a mapping (dict-like) object. If a string then assume it's an fsspec-style URL and load via fsspec - this gives the ability to read from cloud object stores. If it's a mapping then treat it as a store - this gives full flexibility to pass in [any type of zarr store](https://zarr.readthedocs.io/en/stable/api/storage.html).

Second suggestion is to expose a `consolidated` argument, which allows the user to specify if [consolidated metadata](https://zarr.readthedocs.io/en/stable/tutorial.html#consolidating-metadata) has been used (this is an optimisation for cloud stores).